### PR TITLE
feat(data-pipeline): add CLAP audio-embedding producer and read seam

### PR DESCRIPTION
**Pull request by ktinubu**
## Summary

Adds **LAION CLAP audio-embedding conditioning** to the dataset/dataloader path (steps 1–4 of the plan; model-side consumption is a follow-up).

- **Producer** `src/synth_setter/pipeline/data/add_clap.py`: a post-finalization CLI that reads each split file's virtual `audio`, downmixes to mono, runs an injected CLAP encoder, and writes a real `clap` `(N, D)` dataset into the same split file. The functional core takes an injected `encode` callable (fully tested without a model download); `load_clap_audio_encoder` is the thin shell that builds the real encoder via a lazy `transformers` import.
- **Datamodule** `surge_datamodule.py`: `read_clap` flag through `SurgeXTDataset` and a `conditioning="clap"` route in `SurgeDataModule`, mirroring `read_m2l`.
- **Shared constant** `CLAP_EMBEDDING_DIM` single-sourced in `data/vst/shapes.py`; `transformers` added to the `torch` dependency group.

### Why split files, not shards
The resharder's virtual layout only carries the core `DATASET_FIELD_DTYPES`, so a side embedding must live in the split file (`train/val/test.h5`) the datamodule opens directly.

### Robustness
- **Idempotent / crash-safe**: a `clap_complete` marker attr is set only after every row is written; a present-but-incomplete field (crashed run) is recomputed, a completed field is skipped, `--overwrite` forces recompute.
- **Width guard**: `add_clap_embeddings(expected_dim=…)` asserts the encoder width before writing; `main()` pins it to `CLAP_EMBEDDING_DIM` only for the default checkpoint, leaving alternates free to project their own width.

### Deviations from the issue's written scope (by design)
- Uses the HF `transformers` `ClapModel` rather than `laion_clap` (decided with the requester — same LAION weights, cleaner checkpoint pinning).
- HDF5 only; the webdataset path is deferred.
- Includes the datamodule read seam, which #1570 lists as a follow-up — it's the natural consumer of the new field.

### Known follow-ups (advisory, not blocking)
- Model-side `conditioning="clap"` consumption (`SurgeFlowMatchingModule` + a CLAP encoder config) — step 5, separate PR.
- Like `m2l`, `clap` is **not** threaded through `_hungarian_match`, so `ot=True` + `conditioning="clap"` would misalign the embedding from its permuted params. The `m2l` half of this is fixed in #1573 (branch `fix/ot-thread-side-embeddings`); `clap` needs the same one-line follower addition once both land.
- CLI `main()` / real-encoder shell are uncovered (model-download trade-off). Manual e2e: `python -m synth_setter.pipeline.data.add_clap <dataset_dir>` against a small finalized dataset, then train with `datamodule.conditioning=clap`.

## Test plan
- `uv run pytest tests/pipeline/data/test_add_clap.py tests/data/test_surge_datamodule.py` → 83 passed.
- `make format` → clean (ruff, ruff-format, pyright, pydoclint, etc.).
- HDF5 core exercised via an injected fake encoder: ragged batches, mono downmix, sample-rate propagation, completion-marker idempotency (skip/recompute/overwrite), and the `expected_dim` width guard (raise + accept).

Refs #1570

### CODE REVIEW

File: src/synth_setter/data/surge_datamodule.py [hunk 1]
```diff
@@ -147,6 +158,7 @@ def __getitem__(self, idx: int | Sequence[int]):
         return dict(
             mel_spec=mel_spec.contiguous() if mel_spec is not None else None,
             m2l=m2l.contiguous() if m2l is not None else None,
+            clap=clap.contiguous() if clap is not None else None,
             params=param_array.contiguous(),
```
[the remark this answers is not in the document]
Comment (ktinubu) [reply src/synth_setter/data/surge_datamodule.py:164]: Fixed in 90df3bc7. Threaded both `clap` and `m2l` through `_hungarian_match(noise, param_array, mel_spec, audio, m2l, clap)` so every conditioning modality rides the same `col_ind` permutation as `params` and stays row-aligned. Added `test_ot_keeps_clap_and_m2l_aligned_with_permuted_params`, which writes a shard with a shared per-row fingerprint and asserts the `clap`/`m2l` rows track the permuted `params` rows (it fails on the pre-fix code).

**Comment by ktinubu** [comment]
## Verification Results

Behavioral verification at `c13c4395`. **The Level 1 end-to-end run loaded the real `laion/clap-htsat-unfused` model and caught three real bugs** in the previously-untested encoder shell (`load_clap_audio_encoder`) that unit tests with a fake encoder could not — all now fixed and pinned by a regression test:

1. transformers v5 `ClapProcessor` takes `audio=` (singular), not the v4 `audios=`.
2. v5's `ClapFeatureExtractor` does **not** resample — it hard-errors unless the input is already 48 kHz; audio is now resampled 44.1→48 kHz via `torchaudio` before encoding.
3. v5 `get_audio_features` returns a `BaseModelOutputWithPooling`; the projected, L2-normalized 512-d embedding is `.pooler_output`.

pyright missed all three because the model/processor are typed `Any` at the dynamic transformers boundary — exactly why an end-to-end check matters here.

- [x] **Real CLI produces 512-d CLAP embeddings and the datamodule surfaces them** *(Level 1 — full producer→datamodule round-trip with the real model; lower levels can't exercise the transformers API)*
  ```text
  $ python /tmp/clap_e2e_final.py   # builds a 3-split dataset, runs `python -m synth_setter.pipeline.data.add_clap <dir>`
  [1] CLI run on 3-split dataset -> exit 0
  [2] each split clap dataset: shape=(2,512) float32, clap_complete=True, L2-unit-norm OK
  [3] datamodule conditioning='clap' read -> batch['clap'] = (2,512) torch.float32
  ```
  The width matches `CLAP_EMBEDDING_DIM` (512), so the default-checkpoint `expected_dim` guard passes on real output, and the embeddings come out L2-normalized.

- [x] **transformers v5 encoder call-contract pinned against regression** *(Level 2 — invokes `load_clap_audio_encoder` with the transformers/torchaudio boundary mocked; runs in offline CI without a model download)*
  ```text
  $ uv run pytest tests/pipeline/data/test_add_clap.py::test_load_clap_audio_encoder_resamples_to_48k_and_returns_pooler_output -q
  1 passed
  ```
  Asserts the `audio=` kwarg, the 44.1→48 kHz resample call, and `pooler_output` extraction — the exact three contracts the bugs above violated.

- [x] **Idempotency: a completed `clap` field is skipped, an incomplete one is recomputed** *(Level 2 — the `add_clap_embeddings` skip/recompute/overwrite branches are exercised directly against on-disk HDF5; an `_exploding_encode` fake proves the encoder is bypassed on the skip path)*
  ```text
  $ uv run pytest tests/pipeline/data/test_add_clap.py -k "completed or incomplete or overwrite or marks_field" -q
  4 passed
  ```
  CLI-level idempotency was also observed exit-0 on a back-to-back rerun; note that repeated real-model CLI invocations ping the HF Hub and can transiently rate-limit when `HF_TOKEN` is unset (operational, not a code defect).

- [ ] **Large-scale CLI run over a finalized R2 dataset** *(SKIP: needs a real finalized dataset + GPU; not available here. To run: `python -m synth_setter.pipeline.data.add_clap <dataset_root>` against a finalized dataset dir holding `train/val/test.h5`, then train with `datamodule.conditioning=clap`.)*

CI (`make test-fast`, `make format`) covers the unit suite and lint and is not duplicated here.

File: src/synth_setter/pipeline/data/add_clap.py [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+            mono = _downmix_to_mono(audio_ds[start:end])
+            embeddings = np.asarray(encode(mono, sample_rate))
+            if out_ds is None:
+                width = embeddings.shape[1]
+                if expected_dim is not None and width != expected_dim:
+                    raise ValueError(
+                        f"CLAP encoder produced width {width}, expected {expected_dim}"
+                    )
+                out_ds = f.create_dataset(field, shape=(num_rows, width), dtype=np.float32)
+            out_ds[start:end] = embeddings
```
[the remark this answers is not in the document]
Comment (ktinubu) [reply src/synth_setter/pipeline/data/add_clap.py:121]: Fixed in 1d6dde53 — the `clap` dataset is now created with `compression=hdf5plugin.Blosc2()`, matching the core audio/mel/param datasets, so large split files stay compressed on disk.

File: tests/data/test_surge_datamodule.py [hunk 3]
```diff
@@ -616,16 +665,78 @@ def test_ot_with_disabled_modalities_passes_none_through(self, single_h5: Path)
...
+        torch.manual_seed(0)
+        item = dataset[0]
+        param_ids = _unwrap(item["params"])[:, 0]
+        # A real permutation must have occurred, else the test can't detect misalignment.
+        assert not torch.equal(param_ids, torch.from_numpy(row_ids))
+        assert torch.equal(_unwrap(item["clap"])[:, 0], param_ids)
+        assert torch.equal(_unwrap(item["m2l"])[:, 0, 0], param_ids)
 
     def test_returned_dict_always_exposes_full_key_set(self, single_h5: Path) -> None:
         """Every ``__getitem__`` return dict exposes the same five-key contract.
```
[the remark this answers is not in the document]
Comment (ktinubu) [reply tests/data/test_surge_datamodule.py]: Fixed in 1d6dde53 — updated both stale "five-key" docstrings (lines 340 and 742) to "six-key" now that `clap` is part of the `_ALL_TENSOR_KEYS` return contract.
